lib/shopify-cli/contextual_command.rb

```diff
@@ -0,0 +1,37 @@
+require 'shopify_cli'
+
+module ShopifyCli
+  class ContextualCommand < ShopifyCli::Command
+    class << self
+      def available?
+        return true if !Project.directory(Dir.pwd) && available.include?(:top_level)
+        return false unless Project.directory(Dir.pwd)
+        available.include?(Project.current.config['project_type'])
+      end
+
+      def available_in(identifier)
+        available << identifier
+      end
+
+      def app_type?
+        Project.directory(Dir.pwd) && app_type_lookup[self]
+      end
+
+      def app_type(identifier, const, path)
+        autoload(const, "shopify-cli/commands/#{path}") if path
+        app_type_lookup[self] ||= {}
+        app_type_lookup[self][identifier] = const_get(const)
+      end
+
+      def app_type_lookup
+        @app_type_lookup ||= {}
+      end
```

Comment:
This doesn't need to be used necessarily. I moved it here to avoid having multiple classes to inherit behaviour from (`AppTypeCommand` _and_ `ContextualCommand`) or having a module that you need to `include` for commands that leverage `app_type`.

Maybe we need to name this more generically. The intention behind #403 that implemented this is just to allow us to differ the implementation of a command based on the AppType, but this isn't necessarily tied to "Apps" per se, this pattern can be used by any type of project.

Reply:
Oops, I meant #406. My point is it could be a generic feature and then we don't need to have two different `TyperCommand` classes.